C4717 warning

See compile warning: `application.cpp(82): warning C4717: 'app::Application::findDefinition': recursive on all control paths, function will cause runtime stack overflow`

Try replacing recursive call with do while loop here
https://github.com/martymcflywa/ScrabbleDictionary/blob/79371f11d57223d01e075ae6525ec12c7376f6d2/app/Application.cpp#L69

Might as well try replacing stacked `if` statements with a do while here too
https://github.com/martymcflywa/ScrabbleDictionary/blob/79371f11d57223d01e075ae6525ec12c7376f6d2/app/Application.cpp#L39

--- a/app/Application.cpp
+++ b/app/Application.cpp
@@ -9,10 +9,12 @@ using namespace cli;
 using namespace app;
 
 /**
-* \brief Instantiate all the interface implementations, inject dependencies where needed, then construct the Dictionary.
+* \brief Instantiate all the interface implementations, inject dependencies where needed, 
+* then construct the Dictionary.
 */
 Application::Application() :
-    _filepath(string(".\\dictionary2018.txt")), // Switched to hardcoded path, replaces FileResolver::getFilepath()
+    // Switched to hardcoded path, replaces FileResolver::getFilepath()
+    _filepath(string(".\\dictionary2018.txt")),
     _loader(TextFileLoader(_filepath)),
     _formatter(DefinitionFormatter()),
     _printer(DefinitionPrinter(_formatter)),
@@ -35,19 +37,29 @@ void Application::init()
 
 /**
 * \brief Show the main menu. Ask user to select from menu, then perform selected task.
+* If input is invalid, menu is shown again.
+* Retry until user picks correct input or rage quits with CTRL+C.
 */
 void Application::mainMenu()
 {
-    const auto selection = Menu::select();
+    string selection;
 
+    // handy input validation from module 5 loop workshop
+    do
+    {
+        selection = Menu::select();
+    } 
+    while (!isValidMenuItem(selection));
+
+    // bust out of the loop when user picks valid input and do stuff
     if (selection == MenuItem::SEARCH_DEFINITION)
     {
         findDefinition();
         return;
     }
-    if (selection == MenuItem::LONGEST_WORD)
+    if (selection == MenuItem::LONGEST_WORDS)
     {
-        longestWord();
+        longestWords();
         return;
     }
     if (selection == MenuItem::WORDS_END_WITH_LOGY)
@@ -60,32 +72,58 @@ void Application::mainMenu()
         Logger::log(Info, "Shutting down, goodbye");
         exit(0);
     }
-    // tried throwing here but couldn't call back to mainMenu() for retry, stack unwinds on throw,
-    // see http://www.acodersjourney.com/2016/08/top-15-c-exception-handling-mistakes-avoid/
-    Logger::log(Error, selection + " is not a valid selection, try again");
+}
+
+/**
+* \brief Returns true if user input at main menu matches any MenuItem.
+* \param selection The input from the user at main menu.
+* \returns true if input matches any MenuItem.
+*/
+bool Application::isValidMenuItem(const string& selection)
+{
+    const auto isValid = selection == MenuItem::SEARCH_DEFINITION
+        || selection == MenuItem::LONGEST_WORDS
+        || selection == MenuItem::WORDS_END_WITH_LOGY
+        || selection == MenuItem::QUIT;
+
+    if(!isValid)
+        Logger::log(Error, selection + " is not a valid selection, try again");
+
+    return isValid;
+}
+
+/**
+* \brief Asks user to input a word. Dictionary will attempt to find it.
+* If found, will print definition, else will notify user word is not found.
+* Repeat until user chooses to go back to main menu.
+*/
+void Application::findDefinition()
+{
+    string word;
+
+    // replaced recursive call with do-while, resolves C4717 compile warning:
+    // recursive on all control paths, function will cause runtime stack overflow
+    do
+    {
+        word = Menu::findDefinition();
+        Logger::log(Output, "Definition for " + word + ":");
+        Logger::log(_dictionary.getDefinition(word));
+    }
+    while (word != MenuItem::BACK);
+
+    // bump back to main menu when we get the BACK command
     mainMenu();
 }
 
-void Application::findDefinition()
-{
-    const auto word = Menu::findDefinition();
-    
-    // give user option to go back to main menu
-    if (word == MenuItem::BACK)
-        mainMenu();
-
-    Logger::log(Output, "Definition for " + word + ":");
-    Logger::log(_dictionary.getDefinition(word));
-
-    // recurse to search for another word
-    findDefinition();
-}
-
-void Application::longestWord()
+/**
+* \brief Will print longest word/s found in dictionary.
+* Then jumps back to main menu.
+*/
+void Application::longestWords()
 {
     Logger::log(Output, "The longest word/s found in the dictionary:");
     
-    auto longest = _dictionary.getLongestWord();
+    auto longest = _dictionary.getLongestWords();
     for (auto const& word : longest)
         Logger::log(word + " (" + to_string(word.length()) + ")");
 

--- a/app/Application.h
+++ b/app/Application.h
@@ -29,11 +29,32 @@ namespace app
         void init();
         /**
          * \brief Show the main menu. Ask user to select from menu, then perform selected task.
+         * If input is invalid, menu is shown again.
+         * Retry until user picks correct input or rage quits with CTRL+C.
          */
         void mainMenu();
     private:
+        /**
+         * \brief Returns true if user input at main menu matches any MenuItem.
+         * \param selection The input from the user at main menu.
+         * \returns true if input matches any MenuItem. 
+         */
+        static bool isValidMenuItem(const string& selection);
+        /**
+         * \brief Asks user to input a word. Dictionary will attempt to find it.
+         * If found, will print definition, else will notify user word is not found.
+         * Repeat until user chooses to go back to main menu.
+         */
         void findDefinition();
-        void longestWord();
+        /**
+         * \brief Will print longest word/s found in dictionary.
+         * Then returns to main menu.
+         */
+        void longestWords();
+        /**
+         * \brief Prints words that end in 'logy' with length seven or less.
+         * Then returns to main menu.
+         */
         void wordsEndWithLogy();
     };
 }

--- a/cli/Menu.cpp
+++ b/cli/Menu.cpp
@@ -13,7 +13,7 @@ string Menu::select()
 
     const auto menu =
         string("[" + MenuItem::SEARCH_DEFINITION + "] Search for a word's definition\n") +
-        string("[" + MenuItem::LONGEST_WORD + "] Find the longest word/s in the dictionary\n") +
+        string("[" + MenuItem::LONGEST_WORDS + "] Find the longest word/s in the dictionary\n") +
         string("[" + MenuItem::WORDS_END_WITH_LOGY + "] Find words that end in 'logy' and have a length of seven or less characters\n") +
         string("["+ MenuItem::QUIT + "] Quit");
 

--- a/cli/MenuItem.cpp
+++ b/cli/MenuItem.cpp
@@ -4,7 +4,7 @@
 using namespace cli;
 
 const string MenuItem::SEARCH_DEFINITION = "1";
-const string MenuItem::LONGEST_WORD = "2";
+const string MenuItem::LONGEST_WORDS = "2";
 const string MenuItem::WORDS_END_WITH_LOGY = "3";
 const string MenuItem::QUIT = "q";
 const string MenuItem::BACK = "<";

--- a/cli/MenuItem.h
+++ b/cli/MenuItem.h
@@ -11,7 +11,7 @@ namespace cli
     struct MenuItem
     {
         static const string SEARCH_DEFINITION;
-        static const string LONGEST_WORD;
+        static const string LONGEST_WORDS;
         static const string WORDS_END_WITH_LOGY;
         static const string QUIT;
         static const string BACK;

--- a/lib/Dictionary.cpp
+++ b/lib/Dictionary.cpp
@@ -41,10 +41,10 @@ string Dictionary::getDefinition(const string& word)
 }
 
 /**
-* \brief Returns the longest word in the dictionary.
-* \returns The longest word in the dictionary.
+* \brief Returns the longest word/s in the dictionary.
+* \returns The longest word/s in the dictionary.
 */
-list<string> Dictionary::getLongestWord() const
+list<string> Dictionary::getLongestWords() const
 {
     return _extractor.getLongestWords();
 }

--- a/lib/Dictionary.h
+++ b/lib/Dictionary.h
@@ -45,7 +45,7 @@ namespace lib {
         * \brief Returns the longest word/s in the dictionary.
         * \returns The longest word/s in the dictionary.
         */
-        list<string> getLongestWord() const;
+        list<string> getLongestWords() const;
         /**
         * \brief Returns words that end in 'logy' that have a length less than or equal to seven.
         * \returns Words that end in 'logy' that have a length less than or equal to seven.

--- a/test/DictionaryTasksTests.cpp
+++ b/test/DictionaryTasksTests.cpp
@@ -111,7 +111,7 @@ namespace dictionaryTasksTests
 
             WHEN("The longest word is requested")
             {
-                const auto actual = dictionary.getLongestWord();
+                const auto actual = dictionary.getLongestWords();
 
                 // delete the test file, keep it out of version control
                 loader.dispose();
@@ -152,7 +152,7 @@ namespace dictionaryTasksTests
 
             WHEN("The longest word is requested")
             {
-                const auto actual = dictionary.getLongestWord();
+                const auto actual = dictionary.getLongestWords();
 
                 // delete the test file, keep it out of version control
                 loader.dispose();

